Client/Frontend/Home/HistoryPanel.swift

```diff
@@ -6,38 +6,70 @@ import UIKit
 
 import Storage
 
-private func getDate(#dayOffset: Int) -> NSDate {
+private func getDate(#dayOffset: Int) -> NSTimeInterval {
     let calendar = NSCalendar(calendarIdentifier: NSGregorianCalendar)!
     let nowComponents = calendar.components(NSCalendarUnit.YearCalendarUnit | NSCalendarUnit.MonthCalendarUnit | NSCalendarUnit.DayCalendarUnit, fromDate: NSDate())
     let today = calendar.dateFromComponents(nowComponents)!
-    return calendar.dateByAddingUnit(NSCalendarUnit.CalendarUnitDay, value: dayOffset, toDate: today, options: nil)!
+    return calendar.dateByAddingUnit(NSCalendarUnit.CalendarUnitDay, value: dayOffset, toDate: today, options: nil)!.timeIntervalSince1970
 }
 
 class HistoryPanel: SiteTableViewController {
-    private let NumSections = 4
-    private var Today = getDate(dayOffset: 0)
-    private var Yesterday = getDate(dayOffset: -1)
-    private var ThisWeek = getDate(dayOffset: -7)
-
-    private var sectionOffsets = [Int: Int]()
+    private var sections = [Cursor]()
 
     override func reloadData() {
+        var Today = getDate(dayOffset: 0)
+        var Yesterday = getDate(dayOffset: -1)
+        var ThisWeek = getDate(dayOffset: -7)
+
+        var finished = 0
+        func finishedFunc(section: Int) -> (data: Cursor) -> () {
+            return { data in
+                self.sections[section] = data
+                finished++
+                // We current reload data as it comes in
+                self.tableView.reloadData()
+                if finished == self.sections.count {
+                    self.refreshControl?.endRefreshing()
+                }
+            }
+        }
+
+        // Prefill sections with some empty data
+        sections.removeAll(keepCapacity: true)
+        for i in 0..<4 {
+            sections.append(Cursor(status: .Success, msg: "Loading"))
+        }
+
         let opts = QueryOptions()
+        let nilTime: NSTimeInterval? = nil
+        opts.filterType = FilterType.DateRange
         opts.sort = .LastVisit
-        profile.history.get(opts, complete: { (data: Cursor) -> Void in
-            self.refreshControl?.endRefreshing()
-            self.sectionOffsets = [Int: Int]()
-            self.data = data
-            self.tableView.reloadData()
-        })
+
+        opts.filter = DateRange(start: nilTime, end: Today)
+        profile.history.get(opts, complete: finishedFunc(0))
+
+        opts.filter = DateRange(start: Today, end: Yesterday)
+        profile.history.get(opts, complete: finishedFunc(1))
+
+        opts.filter = DateRange(start: Yesterday, end: ThisWeek)
+        profile.history.get(opts, complete: finishedFunc(2))
+
+        opts.filter = DateRange(start: ThisWeek, end: nilTime)
+        profile.history.get(opts, complete: finishedFunc(3))
```

Comment:
There is so much hardcoded here. Wouldn't it be nicer to have someting like a `HistorySection` struct or class to keep all the state in? Then you can probably also get rid of the hardcoded numbers, etc.